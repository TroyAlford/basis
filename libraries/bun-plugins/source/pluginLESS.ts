import type { BunPlugin } from 'bun'
import * as less from 'less'
import * as path from 'node:path'

/**
 * A Bun build plugin that compiles SASS/SCSS files into CSS and injects them into the document head.
 * The plugin automatically handles .scss file imports in your JavaScript/TypeScript code by compiling
 * the SASS content and creating a runtime script that injects the compiled CSS into the page.
 * @returns A Bun build plugin that handles SASS/SCSS compilation and injection
 * @example Basic usage in Bun build configuration
 * ```typescript
 * import { build } from "bun";
 * import { pluginLESS } from "./pluginLESS";
 *
 * await build({
 *   entrypoints: ['./src/index.ts'],
 *   plugins: [
 *     pluginLESS()
 *   ]
 * });
 * ```
 * @example Usage in your JavaScript/TypeScript files
 * // Will be compiled and injected at runtime
 * import './styles.less'
 *
 * // Transformed to:
 * <style data-path="path/to/styles.less">
 * // compiled CSS content
 * </style>
 */
export function pluginLESS(): BunPlugin {
  return {
    name: 'less',
    setup({ config: { target = 'browser' }, onLoad }) {
      onLoad({ filter: /\.less$/ }, async args => ({
        contents: target === 'browser' ? `
          const style = document.createElement('style');
          style.dataset.path = ${JSON.stringify(path.relative(process.cwd(), args.path))};
          style.textContent = ${JSON.stringify(less.render(args.path).then(result => result.css))};
          document.head.appendChild(style);
        ` : '',
        loader: 'js',
      }))
    },
  }
}
