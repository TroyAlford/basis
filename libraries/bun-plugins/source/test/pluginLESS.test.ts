import { describe, expect, test } from 'bun:test'
import * as path from 'node:path'
import { pluginLESS } from '../pluginLESS'

const FIXTURE_FILE = path.join(import.meta.dir, 'input.test.less')
const JS_FILEPATH = './input.test.js'

describe('pluginLESS', () => {
  test.each([true, false])('minified: %p', async minify => {
    const build = await Bun.build({
      entrypoints: [FIXTURE_FILE],
      plugins: [pluginLESS({ minify })],
      target: 'browser',
    })

    expect(build.outputs).toHaveLength(1)
    expect(build.outputs.map(file => file.path)).toMatchSnapshot()
    const js = build.outputs.find(file => file.path === JS_FILEPATH)
    expect(await js.text()).toMatchSnapshot()
  })
})
