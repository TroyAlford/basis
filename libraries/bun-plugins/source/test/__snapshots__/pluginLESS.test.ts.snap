// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`pluginLESS minified: true 1`] = `
[
  "./input.test.js",
]
`;

exports[`pluginLESS minified: true 2`] = `
"// libraries/bun-plugins/source/test/input.test.less
var style = document.createElement("style");
style.dataset.path = "libraries/bun-plugins/source/test/input.test.less";
style.textContent = ".parent>.child{background-color:rgba(51,51,51,0.8);padding:16px}";
document.head.appendChild(style);
"
`;

exports[`pluginLESS minified: false 1`] = `
[
  "./input.test.js",
]
`;

exports[`pluginLESS minified: false 2`] = `
"// libraries/bun-plugins/source/test/input.test.less
var style = document.createElement("style");
style.dataset.path = "libraries/bun-plugins/source/test/input.test.less";
style.textContent = \`.parent > .child {
  background-color: rgba(51, 51, 51, 0.8);
  padding: 16px;
}
\`;
document.head.appendChild(style);
"
`;
